HCollections/HListT.fs

```diff
@@ -9,11 +9,17 @@ type HListT<'ts, 'elem> =
     | Empty of 'ts HList * Teq<'ts, unit>
     | Cons of 'ts HList * 'elem list * HListTConsCrate<'ts, 'elem>
 
-and private HListTConsEvaluator<'ts, 'elem, 'ret> =
+and private HListTConsEval<'ts, 'elem, 'ret> =
     abstract member Eval<'t, 'ts2> : HListT<'ts2, 'elem> -> Teq<'ts, 't -> 'ts2> -> 'ret
 
 and private HListTConsCrate<'ts, 'elem> =
-    abstract member Apply<'ret> : HListTConsEvaluator<'ts, 'elem, 'ret> -> 'ret
+    abstract member Apply<'ret> : HListTConsEval<'ts, 'elem, 'ret> -> 'ret
+
+type HListTCons<'ts, 'elem> =
+    abstract Apply<'r> : HListTConsEvaluator<'ts, 'elem, 'r> -> 'r
+
+and HListTConsEvaluator<'ts, 'elem, 'r> =
+    abstract Eval<'t, 'u> : Teq<'ts, 't -> 'u> * 't * 'elem * HListT<'u, 'elem> -> 'r
```

Comment:
Tupling of the arguments to avoid the hell we see when you start recursing

Reply:
Incredible sadness. Consumers of the API (who are not us) have almost certainly not encountered this before, so please put a nice detailed comment about why this looks so sad.

Reply:
OOI is there any dotnet/fsharp issue about this problem? Should we raise one, do we have an easy to demonstrate example of this?

Reply:
As far as I'm aware we haven't raised an issue around this, we've just noticed the issue. I can double check this and if it is indeed the case I can create a small test case and submit an issue.